doc/mps_youtube.config.rst
mps_youtube.config module
=========================

.. automodule:: mps_youtube.config
    :members:
    :undoc-members:
    :show-inheritance:


doc/mps_youtube.commands.config.rst
mps_youtube.commands.config module
==================================

.. automodule:: mps_youtube.commands.config
    :members:
    :undoc-members:
    :show-inheritance:


CHANGELOG
Version 0.2.6
1 January 2016

Features:
- Add browserplay command to play video in browser (rjvani) (#355 #390)
- Make url command support multiple url, and allow loading urls from file
    (paulnicolet) (#90 #391)
- Make youtube id available to download command (ids1024) (#375)
- Make pagination work properly for all lists of songs (local playlists, etc.)
    and make them all support the dump command (ids1024) (#201)
- Show warning when adding duplicate track to playlist (BensonQiu) (#374 #383)

Bug fixes:
- Support playlists longer than 200 items (ids1024) (#192)
- Make dapl download complete playlist (JKatzwinkel) (#294 #325)
- Use pyperclip instead of xerox, fixing clipboard on linux (ids1024) (#311)
- Fix error message on windows when video title has & (ids1024) (#321)
- Run notifier command in background (hrnr) (#339)
- Correctly handle -v option of play (TimoDritschler) (#352)
- Make it possible to use a pager with --help (livingBEEF) (#349)
- Allow mplayer for https streams if version is new enough
    (ids1024 Evidlo paddatrapper) (#315 #336 #317)
- Use mpv.com instead of mpv.exe on windows (ids1024)

Internal API changes:
- Split code into several files (still a work in progress) (ids1024) (#208)
- Use a python decorator to declare commands (ids1024)
- Create a paginatesongs() function for handling the pagination and dump
  command uniformly as mentioned in features (#399) (ids1024)
- Various others

There have been various other small bug fixes and internal API changes. Use
`git log v0.2.5..0.2.6` to see all of them.

-------------------------------------------------------------------------------

Version 0.2.5
1 June 2015

- Support python 3 only (will not run with python 2)
- Fix progress bar bug on Windows (#215, #223)
- Fix bug in playlist search (#282)
- Add user_order setting for separate order of user searches (lol768) (#277)
- Add mix command (lol768) (#241, #283)
- Add page count with total number of pages and current page (JKatzwinkel)
- Fix UnicodeEncodeError on non-unicode terminals
- Install .desktop in setup.py so it does not have to be installed manually
    required for Ubuntu sound menu support

-------------------------------------------------------------------------------

Version 0.2.4
13 May 2015

 - Fix issue with rate-limiting in "album" command (#217)
 - Implement support for MPRIS (hrnr and ids1024) (#191, #210)
 - Fix display issue with small terminal width (#215)
 - Use version 3 of the youtube api (JKatzwinkel, ids1024, and lol768) (#256)
     version 2 was disabled by youtube and no longer works

-------------------------------------------------------------------------------

Version 0.2.3
17 February 2015

 - Fix issue with opening old playlist file (#214)
 - Fix issue with "set" command
 - Metadata tagging of downloaded audio files
     requires ffmpeg or avconv
     early implementation, uses YouTube video title, needs more work

-------------------------------------------------------------------------------

Version 0.2.2
14 February 2015

- Implement skip to previous track (ids1024) #175 #18 #14
    Use > and < keys for next and prev track, q for quit (was ctrl-c)
    mpv / mplayer input.conf can be imported to use custom keymappings

- Implement optional external download application (ids1024 and np1) #198 #132
    enter "help dl-command" for more info

- Show stream quality on status line (Brebiche38) #134 #163

- Implement support for creating Windows builds with py2exe (ids1024) #190
    Use setup.py in py2exe branch

- Fix progress bar for mpv 0.7 (ids1024) #155 #161
    Resolves issues with playback looping

- Fix mps-youtube does not run on FreeBSD (nivit)

- Fix YouTube comments overshoot window size

- Fix issue with audio remux when using `d` download command

-------------------------------------------------------------------------------

Version 0.2.1
27 November 2014

[Bugfix] - Detect installed player on installation (#149)

-------------------------------------------------------------------------------

Version 0.2.0
25 November 2014

- Auto detect terminal window size

- Transcode audio to MP3 and other formats (requires ffmpeg or avconv)

- Enabled da and dv commands for playlist search results

- Added dapl and dvpl to download YouTube playlist audio/video
    by url/id to a separate subdirectory (mtahmed) #110

- Progress indicator now works with mpv (previously only mplayer)

- Added option to show system notifications (Alex Nisnevich) #95
    (can be used with libnotify - notify-send on linux)

- Added overwrite true/false option for downloads (mtahmed) #93
    (skips download if downloaded file already exists)

- Added copy to clipboard feature #97
    (requires python xerox module and xclip on linux or pywin32 on windows)

- Remux audio downloads for better file compatibility (#135)

[Bugfix] - Accept two-character length usernames (#99)

[Bugfix] - Expand ~ character in download directory setting (punchagan)

[Bugfix] - Don't truncate filename (YoussF) #122

[Bugfix] - Exit with correct status code

-------------------------------------------------------------------------------

Version 0.01.46
May 18 2014

 - Added c <number> to view comments for a video
    (first 50 comments, no reply-comments)

 - Added feature to match album tracks using MusicBrainz (vixus0)
    To search albums, enter "album" optionally followed by album title

 - Custom formatted search result list using "set columns" command
   Optionally shows: rating, likes, dislikes, views, user, date, category
   and comments (number of) in search results

 - Added "set order <relevance|views|rating|date>" command for
     specifying search result ordering

 - Added "set console-width" for setting output width (default 80)

 - Added uploaded date in video info display (request #64)

 - Added likes / dislikes in video info display

-------------------------------------------------------------------------------

Version 0.01.44
Mar 27 2014


mps_youtube/config.py
import os
import re
import sys
import copy
import pickle
from urllib.request import urlopen
from urllib.error import HTTPError
from urllib.parse import urlencode

import pafy

from . import g, c, paths, util


mswin = os.name == "nt"

class ConfigItem(object):

    """ A configuration item. """

    def __init__(self, name, value, minval=None, maxval=None, check_fn=None,
            require_known_player=False, allowed_values=None):
        """ If specified, the check_fn should return a dict.

        {valid: bool, message: success/fail mesage, value: value to set}

        """
        self.default = self.value = value
        self.name = name
        self.type = type(value)
        self.maxval, self.minval = maxval, minval
        self.check_fn = check_fn
        self.require_known_player = require_known_player
        self.allowed_values = []
        if allowed_values:
            self.allowed_values = allowed_values

    @property
    def get(self):
        """ Return value. """
        return self.value

    @property
    def display(self):
        """ Return value in a format suitable for display. """
        retval = self.value

        if self.name == "max_res":
            retval = str(retval) + "p"

        if self.name == "encoder":
            retval = str(retval) + " [%s]" % (str(g.encoders[retval]['name']))

        return retval

    def set(self, value):
        """ Set value with checks. """
        # note: fail_msg should contain %s %s for self.name, value
        #       success_msg should not
        # pylint: disable=R0912
        # too many branches

        success_msg = fail_msg = ""
        value = value.strip()
        value_orig = value

        # handle known player not set

        if self.allowed_values and value not in self.allowed_values:
            fail_msg = "%s must be one of * - not %s"
            allowed_values = copy.copy(self.allowed_values)
            if '' in allowed_values:
                allowed_values[allowed_values.index('')] = "<nothing>"
            fail_msg = fail_msg.replace("*", ", ".join(allowed_values))

        if self.require_known_player and \
                not util.is_known_player(Config.PLAYER.get):
            fail_msg = "%s requires mpv or mplayer, can't set to %s"

        # handle true / false values

        elif self.type == bool:

            if value.upper() in "0 OFF NO DISABLED FALSE".split():
                value = False
                success_msg = "%s set to False" % c.c("g", self.name)

            elif value.upper() in "1 ON YES ENABLED TRUE".split():
                value = True
                success_msg = "%s set to True" % c.c("g", self.name)

            else:
                fail_msg = "%s requires True/False, got %s"

        # handle int values

        elif self.type == int:

            if not value.isdigit():
                fail_msg = "%s requires a number, got %s"

            else:
                value = int(value)

                if self.maxval and self.minval:

                    if not self.minval <= value <= self.maxval:
                        m = " must be between %s and %s, got "
                        m = m % (self.minval, self.maxval)
                        fail_msg = "%s" + m + "%s"

                if not fail_msg:
                    dispval = value or "None"
                    success_msg = "%s set to %s" % (c.c("g", self.name),
                                                    dispval)

        # handle space separated list

        elif self.type == list:
            success_msg = "%s set to %s" % (c.c("g", self.name), value)
            value = value.split()

        # handle string values

        elif self.type == str:
            dispval = value or "None"
            success_msg = "%s set to %s" % (c.c("g", self.name),
                                            c.c("g", dispval))

        # handle failure

        if fail_msg:
            failed_val = value_orig.strip() or "<nothing>"
            colvals = c.y + self.name + c.w, c.y + failed_val + c.w
            return fail_msg % colvals

        elif self.check_fn:
            checked = self.check_fn(value)
            value = checked.get("value") or value

            if checked['valid']:
                value = checked.get("value", value)
                self.value = value
                Config.save()
                return checked.get("message", success_msg)

            else:
                return checked.get('message', fail_msg)

        elif success_msg:
            self.value = value
            Config.save()
            return success_msg


def check_console_width(val):
    """ Show ruler to check console width. """
    valid = True
    message = "-" * val + "\n"
    message += "console_width set to %s, try a lower value if above line ove"\
        "rlaps" % val
    return dict(valid=valid, message=message)


def check_api_key(key):
    """ Validate an API key by calling an API endpoint with no quota cost """
    url = "https://www.googleapis.com/youtube/v3/i18nLanguages"
    query = {"part": "snippet", "fields": "items/id", "key": key}
    try:
        urlopen(url + "?" + urlencode(query)).read()
        message = "The key, '" + key + "' will now be used for API requests."

        # Make pafy use the same api key
        pafy.set_api_key(Config.API_KEY.get)

        return dict(valid=True, message=message)
    except HTTPError:
        message = "Invalid key or quota exceeded, '" + key + "'"
        return dict(valid=False, message=message)


def check_ddir(d):
    """ Check whether dir is a valid directory. """
    expanded = os.path.expanduser(d)
    if os.path.isdir(expanded):
        message = "Downloads will be saved to " + c.y + d + c.w
        return dict(valid=True, message=message, value=expanded)

    else:
        message = "Not a valid directory: " + c.r + d + c.w
        return dict(valid=False, message=message)


def check_win_pos(pos):
    """ Check window position input. """
    if not pos.strip():
        return dict(valid=True, message="Window position not set (default)")

    pos = pos.lower()
    reg = r"(TOP|BOTTOM).?(LEFT|RIGHT)"

    if not re.match(reg, pos, re.I):
        msg = "Try something like top-left or bottom-right (or default)"
        return dict(valid=False, message=msg)

    else:
        p = re.match(reg, pos, re.I).groups()
        p = "%s-%s" % p
        msg = "Window position set to %s" % p
        return dict(valid=True, message=msg, value=p)


def check_win_size(size):
    """ Check window size input. """
    if not size.strip():
        return dict(valid=True, message="Window size not set (default)")

    size = size.lower()
    reg = r"\d{1,4}x\d{1,4}"

    if not re.match(reg, size, re.I):
        msg = "Try something like 720x480"
        return dict(valid=False, message=msg)

    else:
        return dict(valid=True, value=size)


def check_encoder(option):
    """ Check encoder value is acceptable. """
    encs = g.encoders

    if option >= len(encs):
        message = "%s%s%s is too high, type %sencoders%s to see valid values"
        message = message % (c.y, option, c.w, c.g, c.w)
        return dict(valid=False, message=message)

    else:
        message = "Encoder set to %s%s%s"
        message = message % (c.y, encs[option]['name'], c.w)
        return dict(valid=True, message=message)


def check_player(player):
    """ Check player exefile exists and get mpv version. """
    if util.has_exefile(player):
        util.load_player_info(player)

        if "mpv" in player:
            version = "%s.%s.%s" % g.mpv_version
            fmt = c.g, c.w, c.g, c.w, version
            msg = "%splayer%s set to %smpv%s (version %s)" % fmt
            return dict(valid=True, message=msg, value=player)

        else:
            msg = "%splayer%s set to %s%s%s" % (c.g, c.w, c.g, player, c.w)
            return dict(valid=True, message=msg, value=player)

    else:
        if mswin and not (player.endswith(".exe") or player.endswith(".com")):
            # Using mpv.exe has issues; use mpv.com
            if "mpv" in player:
                retval = check_player(player + ".com")
                if retval["valid"]:
                    return retval
            return check_player(player + ".exe")

        else:
            msg = "Player application %s%s%s not found" % (c.r, player, c.w)
            return dict(valid=False, message=msg)


class _Config(object):

    """ Holds various configuration values. """

    _configitems = [
            ConfigItem("order", "relevance",
                allowed_values="relevance date views rating title".split()),
            ConfigItem("user_order", "", allowed_values =
                [""] + "relevance date views rating".split()),
            ConfigItem("max_results", 19, maxval=50, minval=1),
            ConfigItem("console_width", 80, minval=70,
                maxval=880, check_fn=check_console_width),
            ConfigItem("max_res", 2160, minval=192, maxval=2160),
            ConfigItem("player", "mplayer" + ".exe" * mswin,
                check_fn=check_player),
            ConfigItem("playerargs", ""),
            ConfigItem("encoder", 0, minval=0, check_fn=check_encoder),
            ConfigItem("notifier", ""),
            ConfigItem("checkupdate", True),
            ConfigItem("show_mplayer_keys", True, require_known_player=True),
            ConfigItem("fullscreen", False, require_known_player=True),
            ConfigItem("show_status", True),
            ConfigItem("columns", ""),
            ConfigItem("ddir", paths.get_default_ddir(), check_fn=check_ddir),
            ConfigItem("overwrite", True),
            ConfigItem("show_video", False),
            ConfigItem("search_music", True),
            ConfigItem("window_pos", "", check_fn=check_win_pos,
                require_known_player=True),
            ConfigItem("window_size", "",
                check_fn=check_win_size, require_known_player=True),
            ConfigItem("download_command", ''),
            ConfigItem("audio_format", "auto",
                allowed_values="auto webm m4a".split()),
            ConfigItem("api_key", "AIzaSyCIM4EzNqi1in22f4Z3Ru3iYvLaY8tc3bo",
                check_fn=check_api_key),
            ConfigItem("set_title", True)
            ] 

    def __getitem__(self, key):
        # TODO: Possibly more efficient algorithm, w/ caching
        for i in self._configitems:
            if i.name.upper() == key:
                return i
        raise KeyError

    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError:
            raise AttributeError

    def __iter__(self):
        return (i.name.upper() for i in self._configitems)

    def save(self):
        """ Save current config to file. """
        config = {setting: self[setting].value for setting in self}

        with open(g.CFFILE, "wb") as cf:
            pickle.dump(config, cf, protocol=2)

        util.dbg(c.p + "Saved config: " + g.CFFILE + c.w)

    def load(self):
        """ Override config if config file exists. """
        if os.path.exists(g.CFFILE):

            with open(g.CFFILE, "rb") as cf:
                saved_config = pickle.load(cf)

            for k, v in saved_config.items():

                try:
                    self[k].value = v

                except KeyError:  # Ignore unrecognised data in config
                    util.dbg("Unrecognised config item: %s", k)

            # Update config files from versions <= 0.01.41
            if isinstance(self.PLAYERARGS.get, list):
                self.WINDOW_POS.value = "top-right"
                redundant = ("-really-quiet --really-quiet -prefer-ipv4 -nolirc "
                             "-fs --fs".split())

                for r in redundant:
                    util.dbg("removing redundant arg %s", r)
                    util.list_update(r, self.PLAYERARGS.value, remove=True)

                self.PLAYERARGS.value = " ".join(self.PLAYERARGS.get)
                self.save()

Config = _Config()
del _Config # _Config is a singleton and should not have more instances
# Prevent module from being deleted
# http://stackoverflow.com/questions/5365562/why-is-the-value-of-name-changing-after-assignment-to-sys-modules-name
ref = sys.modules[__name__]
# Any module trying to import config will get the Config object instead
sys.modules[__name__] = Config
